src/consumer/stream_consumer.rs

```diff
@@ -79,65 +71,65 @@ unsafe impl Send for PolledMessagePtr {}
 /// advised to use automatic commit, as some messages might have been consumed by the internal Kafka
 /// consumer but not processed. Manual offset storing should be used, see the `store_offset`
 /// function on `Consumer`.
+#[pin_project]
 pub struct MessageStream<'a, C: ConsumerContext + 'static> {
     consumer: &'a StreamConsumer<C>,
-    receiver: mpsc::Receiver<Option<PolledMessagePtr>>,
+    poll_interval: Duration,
+    send_none: bool,
+    should_stop: Arc<AtomicBool>,
 }
 
 impl<'a, C: ConsumerContext + 'static> MessageStream<'a, C> {
     fn new(
         consumer: &'a StreamConsumer<C>,
-        receiver: mpsc::Receiver<Option<PolledMessagePtr>>,
-    ) -> MessageStream<'a, C> {
-        MessageStream { consumer, receiver }
+        poll_interval: Duration,
+        send_none: bool,
+        should_stop: Arc<AtomicBool>,
+    ) -> Self {
+        Self {
+            consumer: consumer,
+            poll_interval: poll_interval,
+            send_none: send_none,
+            should_stop: should_stop,
+        }
     }
 }
 
-impl<'a, C: ConsumerContext + 'a> Stream for MessageStream<'a, C> {
+impl<'a, C: ConsumerContext> Stream for MessageStream<'a, C> {
     type Item = KafkaResult<BorrowedMessage<'a>>;
 
-    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
-        self.receiver.poll_next_unpin(cx).map(|ready| {
-            ready.map(|polled_ptr_opt| {
-                polled_ptr_opt.map_or(Err(KafkaError::NoMessageReceived), |polled_ptr| {
-                    polled_ptr.into_message_of(self.consumer)
-                })
-            })
-        })
-    }
-}
-
-/// Internal consumer loop. This is the main body of the thread that will drive the stream consumer.
-/// If `send_none` is true, the loop will send a None into the sender every time the poll times out.
-fn poll_loop<C: ConsumerContext>(
-    consumer: &BaseConsumer<C>,
-    mut sender: mpsc::Sender<Option<PolledMessagePtr>>,
-    should_stop: &AtomicBool,
-    poll_interval: Duration,
-    send_none: bool,
-) {
-    trace!("Polling thread loop started");
-    while !should_stop.load(Ordering::Relaxed) {
-        trace!("Polling base consumer");
-        let future_sender = match consumer.poll_raw(Timeout::After(poll_interval)) {
-            None => {
-                if send_none {
-                    sender.send(None)
-                } else {
-                    continue; // TODO: check stream closed
+    fn poll_next(self: Pin<&mut Self>, _cx: &mut Context) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+        debug!("Polling stream for next message");
+        loop {
+            if this.should_stop.load(Ordering::Relaxed) {
+                info!("Stopping");
+                return Poll::Ready(None);
+            } else {
+                let timeout = crate::util::Timeout::After(*this.poll_interval);
+                let res = this.consumer.consumer.poll(timeout);
```

Comment:
This seems quite problematic! If someone requests a large timeout they'll starve out all other futures on this core, and if they request a small timeout they'll spin hot. To put it another way, this future never returns `Poll::Pending`, and therefore is not going to play nice with other futures. My intuition suggests that the old thread-based code will be far more performant here for many workloads.

Reply:
@benesch , what if we use `message_queue_nonempty_callback` to wake this stream? Then we can do `consumer.poll(0)`